### PR TITLE
feat: custom tools — inject user-defined shell-command tools

### PR DESCRIPTION
**Pull request by y0usaf**
## Summary

- Adds a \`customTools\` setting that injects user-defined, shell-backed tools into Claude Code's tool list
- Uses CC's own \`buildTool()\` factory (detected via regex in the minified bundle) so custom tools are constructed exactly the same way as every native CC tool — inheriting all default method implementations automatically
- Renders tool-use and results with \`React.createElement\` using detected Text/Box component variables, matching the rendering approach of every other CC tool
- \`checkPermissions\` delegates to BashTool so custom tool commands match the user's existing Bash allow/deny rules (e.g. \`Bash(git *)\`)
- \`validateInput\` type-checks declared parameters before execution
- Two injection strategies: appends after the toolsets filter (Strategy A) or spreads into the original aggregation (Strategy B)
- 27 tests covering helper detection, injection strategies, collision guard, rendering, and edge cases

## New helpers

- \`findBuildToolFunc\`: detects \`function NAME(PARAM){return{...DEFAULTS,userFacingName:()=>PARAM.name,...PARAM}}\` in the minified bundle
- \`getCwdFuncName\`: detects CC's three-step \`getCwdState → pwd → getCwd\` chain for correct session working directory

## Config shape

```json
{
  "settings": {
    "customTools": [
      {
        "name": "DateTool",
        "description": "Get the current date and time",
        "parameters": {},
        "command": "date"
      },
      {
        "name": "WeatherTool",
        "description": "Get weather for a city",
        "parameters": {
          "city": { "type": "string", "description": "City name" }
        },
        "command": "curl -s wttr.in/{{city}}?format=3",
        "shell": "bash",
        "timeout": 10000
      }
    ]
  }
}
```

Optional fields: `shell` (default `"sh"`), `timeout` (default `30000` ms), `workingDir`, `env` (extra env vars), `prompt` (override the auto-generated LLM description).

## Test plan

- [ ] Apply with no `customTools` — patch skips cleanly, no change to cli.js
- [ ] Apply with one tool, no toolsets — Strategy B fires, tool appears in Claude's tool list
- [ ] Apply with toolsets + custom tools — Strategy A fires, custom tool appears regardless of active toolset
- [ ] Tool with parameters — `{{param}}` placeholders substituted in rendered command and execution
- [ ] Tool that exits non-zero — stderr and exit code appear in the result
- [ ] Tool with `prompt` override — custom prompt string is sent to the model
- [ ] Tool name colliding with a built-in — patch returns null with a clear error

## Summary by CodeRabbit

* **New Features**
  * Custom tools: define user tools (name, description, parameters, command, shell, timeout, working dir, env, prompt) and inject them into the toolset.

* **Bug Fixes**
  * Clearer native-installation error messages including native load-failure details and a NixOS/Bun troubleshooting hint when applicable.

* **Improvements**
  * More robust ELF extraction/repacking for native installations; config validation now filters invalid custom-tool entries with warnings.

* **Tests**
  * New tests covering custom-tool injection and native-installation error messaging.

### CODE REVIEW

**Comment by y0usaf** [comment]
@coderabbitai review

File: src/config.ts [hunk 1]
```diff
@@ -219,6 +220,14 @@ const normalizeConfig = (config: TweakccConfig): void => {
     );
   }
 
+  // Merge each custom tool item against the default template
+  if (config.settings.customTools) {
+    config.settings.customTools = config.settings.customTools.map(
+      tool =>
+        deepMergeWithDefaults(tool, DEFAULT_CUSTOM_TOOL) as typeof DEFAULT_CUSTOM_TOOL
+    );
+  }
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/config.ts]: Fixed in fa2bce1 — `config.ts` now checks `Array.isArray`, validates each field's type explicitly (not just truthiness), logs only index/name/invalid keys (not the full tool object), and rejects arrays for `parameters`.

File: src/nativeInstallation.ts [hunk 2]
```diff
@@ -705,9 +784,19 @@ export function extractClaudeJsFromNativeInstallation(
...
+
+    if (isELFFile(nativeInstallationPath)) {
+      ({ bunOffsets, bunData, moduleStructSize } = extractBunDataFromELFRaw(
+        nativeInstallationPath
+      ));
+    } else {
+      LIEF.logging.disable();
+      const binary = LIEF.parse(nativeInstallationPath);
+      ({ bunOffsets, bunData, moduleStructSize } = getBunData(binary));
+    }
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/nativeInstallation.ts:802]: Fixed in 4291e1c — both `extractClaudeJsFromNativeInstallation` and `repackNativeInstallation` now try the raw ELF path first and catch any exception (including 'BUN trailer not found'), falling back to the LIEF path for section-based ELF builds.

File: src/patches/customTools.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+  boxComponent: string,
+  requireFunc: string
+): string => {
+  const nameJson = JSON.stringify(tool.name);
+  const promptString = generatePromptString(tool);
+  const promptJson = JSON.stringify(promptString);
+  const descJson = JSON.stringify(tool.description);
+  const cmdJson = JSON.stringify(tool.command);
+  const shellJson = JSON.stringify(tool.shell ?? 'sh');
+  const timeoutVal = tool.timeout ?? 30000;
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/patches/customTools.ts]: Fixed in 9eeaf82 — the shell basename is now inspected at code-generation time and mapped to its exec flag (`cmd` → `/c`, `powershell`/`pwsh` → `-Command`, everything else → `-c`), so the generated `spawnSync` call uses the correct flag for the configured shell.

File: src/patches/customTools.ts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+  const schemaJson = JSON.stringify({ type: 'object', properties, required });
+
+  // Build parameter substitution code: one replace per param
+  const substitutions = Object.keys(tool.parameters)
+    .map(
+      k =>
+        `cmd=cmd.replace(/\\{\\{${k}\\}\\}/g,String(args[${JSON.stringify(k)}]??""));`
+    )
+    .join('');
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/patches/customTools.ts]: Fixed in 4291e1c — added an `escapeForRegex` helper that escapes regex-special characters before embedding parameter names in regex literals. All three substitution sites (argsSubst, inputSubst, safeInputSubst) now use it, and safeInputSubst uses the same empty-string fallback as the others for consistency.

File: src/patches/customTools.ts [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  }
+
+  for (const tool of customTools) {
+    if (BUILTIN_TOOL_NAMES.has(tool.name)) {
+      console.error(
+        `patch: customTools: tool "${tool.name}" collides with a built-in CC tool name — rename it`
+      );
+      return null;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/patches/customTools.ts:345]: Fixed in 4291e1c — added a `seenToolNames` Set in the validation loop; returns `null` with an error message if any two user-defined tools share a name.

File: src/config.ts [hunk 6]
```diff
@@ -227,6 +227,24 @@ const normalizeConfig = (config: TweakccConfig): void => {
...
+        !tool.description ||
+        !tool.parameters
+      ) {
+        console.warn(
+          `config: customTools: dropping invalid tool (missing required fields): ${JSON.stringify(tool)}`
+        );
+        return false;
+      }
+      return true;
+    });
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/config.ts]: Fixed in fa2bce1 — validation now guards the array container first, checks each field's type explicitly, and logs only index + name + list of invalid keys (no `JSON.stringify` of the whole tool).

File: src/tests/config.test.ts [hunk 7]
```diff
@@ -1523,6 +1524,141 @@ describe('config.ts', () => {
...
+        nativeInstallation,
+        'extractClaudeJsFromNativeInstallation'
+      ).mockResolvedValue(null);
+      vi.spyOn(nativeInstallation, 'getNativeModuleLoadError').mockReturnValue(
+        'libstdc++.so.6: cannot open shared object file: No such file or directory'
+      );
+
+      await expect(
+        findClaudeCodeInstallation(mockConfig, { interactive: true })
+      ).rejects.toThrow('node-lief failed to load');
```
[the remark this answers is not in the document]
Comment (y0usaf) [reply src/tests/config.test.ts]: Fixed in fa2bce1 — assertion updated to `'Reason: node-lief failed to load: libstdc++.so.6'`, verifying the concrete load reason is included.

**Comment by y0usaf** [comment]
Addressed all review comments:

- **Regex escaping**: Added `escapeForRegex` helper; all three substitution sites now escape param names and use consistent empty-string fallback
- **Duplicate tool names**: Added `seenToolNames` Set check in the validation loop
- **ELF raw path**: Both extract and repack functions now try the raw path first, catch any error, and fall back to LIEF for section-based ELF builds
- **Shell exec flag**: Generated `spawnSync` call now maps shell basename to correct flag (`cmd` → `/c`, `powershell`/`pwsh` → `-Command`, others → `-c`)
- **Config validation**: `customTools` validation now guards array container, checks types explicitly, logs index/name/invalid-keys only (no secret leakage), rejects non-object `parameters`
- **Test assertion**: Updated to assert concrete error substring `'Reason: node-lief failed to load: libstdc++.so.6'`
- **New test**: Added generic (non-shared-object) load error test; also fixed the two pre-existing test failures using `vi.spyOn` instead of `vi.mocked`

**Comment by y0usaf** [comment]
@coderabbitai review

**Comment by y0usaf** [comment]
@coderabbitai review

**Comment by y0usaf** [comment]
@coderabbitai review

**Comment by y0usaf** [comment]
@coderabbitai review

**Comment by y0usaf** [comment]
@coderabbitai review

**Comment by y0usaf** [comment]
Closing in favor of clean replacement #711, rebased on current main and stripped of unrelated native/Nix changes.
